File: Python/Next_Permutation.py
# https://leetcode.com/problems/next-permutation/

class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i = len(nums) - 2
        while(i >= 0 and nums[i] >= nums[i + 1]):
            i -= 1
        
        if(i == -1):
            nums.sort()
            return
        
        j = len(nums) - 1
        while(j > i):
            if(nums[j] > nums[i]):
                nums[j], nums[i] = nums[i], nums[j]
                break
            j -= 1
        
        start = i + 1
        end = len(nums) -1
        while(start < end):
            nums[start], nums[end] = nums[end], nums[start]
            start += 1
            end -= 1

File: C++/Next_Permutation.cpp
// https://leetcode.com/problems/next-permutation/
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int g=-1,s=-1,n=nums.size(),i=n-1;
        for(i=n-1;i>0;i--){
            if(nums[i]>nums[i-1]){
                g=s;
                s=i-1;
                break;
            }
        }
        if(s!=-1){
            for(i=n-1;i>g;i--){
                if(nums[i]>nums[s]){
                    g=i;
                    break;
                }
            }
            swap(nums[g],nums[s]);
            reverse(nums.begin()+s+1,nums.end());
        }
        else
            reverse(nums.begin(),nums.end());
    }
};